internal/datastore/postgres/reader.go

```diff
@@ -0,0 +1,206 @@
+func (r *pgReader) ReverseQueryRelationships(
+	ctx context.Context,
+	subjectFilter *v1.SubjectFilter,
+	opts ...options.ReverseQueryOptionsOption,
+) (iter datastore.RelationshipIterator, err error) {
+	qBuilder := common.NewSchemaQueryFilterer(schema, r.filterer(queryTuples)).
+		FilterToSubjectFilter(subjectFilter)
+
+	queryOpts := options.NewReverseQueryOptionsWithOptions(opts...)
+
+	if queryOpts.ResRelation != nil {
+		qBuilder = qBuilder.
+			FilterToResourceType(queryOpts.ResRelation.Namespace).
+			FilterToRelation(queryOpts.ResRelation.Relation)
+	}
+
+	return r.querySplitter.SplitAndExecuteQuery(ctx,
+		qBuilder,
+		options.WithLimit(queryOpts.ReverseLimit),
+	)
+}
+
+func (r *pgReader) ReadNamespace(ctx context.Context, nsName string) (*core.NamespaceDefinition, datastore.Revision, error) {
+	ctx, span := tracer.Start(ctx, "ReadNamespace", trace.WithAttributes(
+		attribute.String("name", nsName),
+	))
+	defer span.End()
+
+	tx, txCleanup, err := r.txSource(ctx)
+	if err != nil {
+		return nil, datastore.NoRevision, fmt.Errorf(errUnableToReadConfig, err)
+	}
+	defer txCleanup(ctx)
+
+	loaded, version, err := loadNamespace(ctx, nsName, tx, r.filterer(readNamespace))
+	switch {
+	case errors.As(err, &datastore.ErrNamespaceNotFound{}):
+		return nil, datastore.NoRevision, err
+	case err == nil:
+		return loaded, version, nil
+	default:
+		return nil, datastore.NoRevision, fmt.Errorf(errUnableToReadConfig, err)
+	}
+}
+
+func loadNamespace(ctx context.Context, namespace string, tx pgx.Tx, baseQuery sq.SelectBuilder) (*core.NamespaceDefinition, datastore.Revision, error) {
+	ctx = datastore.SeparateContextWithTracing(ctx)
+
+	ctx, span := tracer.Start(ctx, "loadNamespace")
+	defer span.End()
+
+	sql, args, err := baseQuery.Where(sq.Eq{colNamespace: namespace}).ToSql()
+	if err != nil {
+		return nil, datastore.NoRevision, err
+	}
+
+	var config []byte
+	var version datastore.Revision
+	err = tx.QueryRow(ctx, sql, args...).Scan(&config, &version)
+	if err != nil {
+		if errors.Is(err, pgx.ErrNoRows) {
+			err = datastore.NewNamespaceNotFoundErr(namespace)
+		}
+		return nil, datastore.NoRevision, err
+	}
+
+	loaded := &core.NamespaceDefinition{}
+	err = proto.Unmarshal(config, loaded)
+	if err != nil {
+		return nil, datastore.NoRevision, err
+	}
+
+	return loaded, version, nil
+}
+
+func (r *pgReader) ListNamespaces(ctx context.Context) ([]*core.NamespaceDefinition, error) {
+	ctx = datastore.SeparateContextWithTracing(ctx)
+
+	tx, txCleanup, err := r.txSource(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer txCleanup(ctx)
+
+	nsDefs, err := loadAllNamespaces(ctx, tx, r.filterer(readNamespace))
+	if err != nil {
+		return nil, fmt.Errorf(errUnableToListNamespaces, err)
+	}
+
+	return nsDefs, err
+}
+
+func loadAllNamespaces(ctx context.Context, tx pgx.Tx, query sq.SelectBuilder) ([]*core.NamespaceDefinition, error) {
+	sql, args, err := query.ToSql()
+	if err != nil {
+		return nil, err
+	}
+
+	var nsDefs []*core.NamespaceDefinition
```

Comment:
move this to right above `for rows.Next()` so it's clear that's what the loop is doing

Reply:
Pre-allocate it to what? We have no idea how many rows we're going to get back.